File: exercises/2η Σειρά/save the cat/save_the_cat.py
from collections import deque
import sys


file = open(sys.argv[1], "rt")
a=[]
N=1
M=0
x=0
y=0
flag=False
queue = deque([])
while True :
	c = file.read(1)
	if not c:
		N=N-1
		break
	if c=="\n":
		flag=True
	if flag==False: 
		M=M+1
	if c=="\n":
		N=N+1	
		x=x+1
	elif c=="W":
		a.append([x%N,y%M,-1,True,"","w"])
		queue.append((x%N,y%M,-1))
		y=y+1
	elif c=="A":
		a.append([x%N,y%M,-1,True,"","a"])
		cat=[x%N,y%M,0]
		y=y+1
	elif c==".":
		a.append([x%N,y%M,-1,True,"","."])
		y=y+1
	else:
		a.append([x%N,y%M,-9,True,"","x"])
		y=y+1

while True:
	if queue:
		value = queue.popleft()
		if a[value[0]*M + value[1]][2]==-1:		
			x=value[0]
			y=value[1]
			t=value[2]
			t=t+1
			a[x*M + y][2]=t
			if x>0:
				x=x-1
				if a[x*M+y][5]!="x":
					queue.append((x,y,t))
				x=x+1
			if x<N-1:
				x=x+1
				if a[x*M+y][5]!="x":
					queue.append((x,y,t))
				x=x-1
			if y>0:
				y=y-1			
				if a[x*M+y][5]!="x":
					queue.append((x,y,t))
				y=y+1
			if y<M-1:
				y=y+1			
				if a[x*M+y][5]!="x":
					queue.append((x,y,t))
				y=y-1
	else:
		break



queue.append([cat[0],cat[1],0])
max_time=-2
b=cat
while queue:
	cat_new=queue.popleft()
	safe_time=a[cat_new[0]*M+cat_new[1]][2]-1
	if(safe_time>max_time):
		max_time=safe_time
		cat=cat_new
	elif(safe_time==max_time):
		if(cat_new[0]<cat[0] or (cat_new[0]==cat[0] and cat_new[1]<cat[1])):
			cat=cat_new
	cat_new[2]=cat_new[2]+1
	if(cat_new[0]<N-1):
		cat_new[0]=cat_new[0]+1
		if(a[cat_new[0]*M+cat_new[1]][2]>cat_new[2] and a[cat_new[0]*M+cat_new[1]][3]):
			a[cat_new[0]*M+cat_new[1]][3]=False
			a[cat_new[0]*M+cat_new[1]][4]=a[(cat_new[0]-1)*M+cat_new[1]][4]+"D"
			queue.append([cat_new[0],cat_new[1],cat_new[2]])
		cat_new[0]=cat_new[0]-1
	if(cat_new[1]>0):
		cat_new[1]=cat_new[1]-1
		if((a[cat_new[0]*M+cat_new[1]][2]>cat_new[2] or a[cat_new[0]*M+cat_new[1]][2]==-1) and a[cat_new[0]*M+cat_new[1]][3]):
			a[cat_new[0]*M+cat_new[1]][3]=False
			a[cat_new[0]*M+cat_new[1]][4]=a[(cat_new[0])*M+cat_new[1]+1][4]+"L"
			queue.append([cat_new[0],cat_new[1],cat_new[2]])
		cat_new[1]=cat_new[1]+1
	if(cat_new[1]<M-1):
		cat_new[1]=cat_new[1]+1
		if((a[cat_new[0]*M+cat_new[1]][2]>cat_new[2] or a[cat_new[0]*M+cat_new[1]][2]==-1) and a[cat_new[0]*M+cat_new[1]][3]):
			a[cat_new[0]*M+cat_new[1]][3]=False
			a[cat_new[0]*M+cat_new[1]][4]=a[(cat_new[0])*M+cat_new[1]-1][4]+"R"
			queue.append([cat_new[0],cat_new[1],cat_new[2]])
		cat_new[1]=cat_new[1]-1
	if(cat_new[0]>0):
		cat_new[0]=cat_new[0]-1
		if((a[cat_new[0]*M+cat_new[1]][2]>cat_new[2] or a[cat_new[0]*M+cat_new[1]][2]==-1) and a[cat_new[0]*M+cat_new[1]][3]):
			a[cat_new[0]*M+cat_new[1]][3]=False
			a[cat_new[0]*M+cat_new[1]][4]=a[(cat_new[0]+1)*M+cat_new[1]][4]+"U"
			queue.append([cat_new[0],cat_new[1],cat_new[2]])
		cat_new[0]=cat_new[0]+1

if(a[cat[0]*M+cat[1]][2]==-1):
	print("infinity")
else:
	print(a[cat[0]*M+cat[1]][2]-1)
if(b[0]==cat[0] and b[1]==cat[1]):
	print("stay")
else:
	print(a[cat[0]*M+cat[1]][4])

File: exercises/3η Σειρά/save_the_cat/SaveTheCat.java
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.LinkedList; 
import java.util.Queue; 
import java.util.List;
//import pos.java;


//http://www.java2s.com/Code/Java/File-Input-Output/Readfilecharacterbycharacter.htm
 class pos {
	int x,y;
	int cat_time=0;
	int bfs_flood=-1;
	int time = -1;
	char sign;
	boolean visited = true;
	String path="";
	
	public pos (int x, int y, char sign){
		this.x = x;
		this.y = y;
		this.sign = sign;
	}
	public pos(){}
	public pos (pos a){
		this.x=a.x;
		this.y=a.y;
		this.time=a.time;
		this.sign=a.sign;
		this.visited=a.visited;
		this.path=a.path;
	}
		
	public String toString(){
		return this.bfs_flood+" ";
	}
}
public class SaveTheCat{
	public static void main (String [] args){
		int N=0,M=0,i=0;
		List<pos> A = new ArrayList<pos>();
		Queue<pos> queue = new LinkedList<>();
		pos cat = new pos();

		File file = new File(args[0]);
    		if (!file.exists()) {
      			System.out.println(args[0] + " does not exist.");
      			return;
    		}
    		if (!(file.isFile() && file.canRead())) {
      			System.out.println(file.getName() + " cannot be read from.");
      			return;
    		}
    		try {
      			FileInputStream fis = new FileInputStream(file);
      			char current;
      			while (fis.available() > 0) {
        			current = (char) fis.read();
        			if(current == '\n'){
					N++;
					M=i;
					i=0;
				}
				else{

					if(current == '.'){
						pos a = new pos(N,i,'.');
						A.add(N*M + i, a);
					}
					else if(current == 'W'){
						pos a = new pos(N,i,'W');
						A.add(N*M + i, a);
 						queue.add(a);
					}
					else if(current == 'X'){
						pos a = new pos(N,i,'X');
						a.time=-9;
						A.add(N*M + i,a);
					}
					else if(current == 'A'){
						cat = new pos(N,i,'A');
						pos a=cat;
						A.add(N*M + i,a);						
					}
					i++;
      				}
    			}
		}
		 catch (IOException e) {
  			e.printStackTrace();
		}
		

		pos a,b;
		int time;
		while(queue.size()!=0){
		a = queue.poll();
		if(A.get(a.x*M + a.y).time==-1){
			time = A.get(a.x*M + a.y).bfs_flood;
			time++;
			A.get(a.x*M + a.y).time = time;	
			if(A.get(a.x*M + a.y).x>0)
				if(A.get((a.x-1)*M + a.y).sign != 'X'){
					b = A.get((a.x-1)*M + a.y);
					b.bfs_flood = time;
					queue.add(b);
				}
			if(A.get(a.x*M + a.y).x<N-1)
				if(A.get((a.x+1)*M + a.y).sign != 'X'){
					b = A.get((a.x+1)*M + a.y);
					b.bfs_flood = time;
					queue.add(b);
				}
			if(A.get(a.x*M + a.y).y>0)
				if(A.get(a.x*M + a.y-1).sign != 'X'){
					b = A.get(a.x*M + a.y-1);
					b.bfs_flood = time;
					queue.add(b);
				}
			if(A.get(a.x*M + a.y).y<M-1)
				if(A.get(a.x*M + a.y+1).sign != 'X'){
					b = A.get(a.x*M + a.y+1);
					b.bfs_flood = time;
					queue.add(b);
				}

	

		}
		}
		/*System.out.println(N +" " +M);
		for(i=0; i<A.size(); i++){
			if(i%M==0) System.out.println();
			if(A.get(i).time==-1) System.out.print(" X "); 
			else System.out.print(A.get(i));		
		}System.out.println();*/
		queue.add(cat);
		int safe_time;
		int max_time = -2;
		int cat_time;
		pos original_cat = new pos(cat);
		while(queue.size()!=0){
			pos cat_new = queue.poll();
			safe_time = A.get(cat_new.x*M + cat_new.y).time-1;
			if(safe_time>max_time){
				max_time=safe_time;
				cat = cat_new;
			}
			else if(safe_time==max_time)
				if(cat_new.x<cat.x || cat_new.x==cat.x && cat_new.y<cat.y)
					cat = cat_new;
			cat_time=cat_new.cat_time+1;
			if(cat_new.x<N-1){
				cat_new.x++;
				if(A.get(cat_new.x*M + cat_new.y).time>cat_new.cat_time && A.get(cat_new.x*M + cat_new.y).visited){
					A.get(cat_new.x*M + cat_new.y).cat_time=cat_time;
					A.get(cat_new.x*M + cat_new.y).visited = false;
					A.get(cat_new.x*M + cat_new.y).path = A.get((cat_new.x-1)*M + cat_new.y).path + "D";
					queue.add(A.get(cat_new.x*M + cat_new.y));
				}
				cat_new.x--;
			}
			if(cat_new.y>0){
				cat_new.y--;
				if((A.get(cat_new.x*M + cat_new.y).time>cat_new.cat_time ||A.get(cat_new.x*M + cat_new.y).time==-1) && A.get(cat_new.x*M + cat_new.y).visited){
					A.get(cat_new.x*M + cat_new.y).cat_time=cat_time;
					A.get(cat_new.x*M + cat_new.y).visited = false;
					A.get(cat_new.x*M + cat_new.y).path = A.get(cat_new.x*M + cat_new.y+1).path+"L";
					queue.add(A.get(cat_new.x*M + cat_new.y));
				}
				cat_new.y++;
			}
			if(cat_new.y<M-1){
				cat_new.y++;
				if((A.get(cat_new.x*M + cat_new.y).time>cat_new.cat_time ||A.get(cat_new.x*M + cat_new.y).time==-1) && A.get(cat_new.x*M + cat_new.y).visited){
					A.get(cat_new.x*M + cat_new.y).cat_time=cat_time;
					A.get(cat_new.x*M + cat_new.y).visited = false;
					A.get(cat_new.x*M + cat_new.y).path = A.get(cat_new.x*M + cat_new.y-1).path+"R";
					queue.add(A.get(cat_new.x*M + cat_new.y));
				}
				cat_new.y--;
			}
			if(cat_new.x>0){
				cat_new.x--;
				if((A.get(cat_new.x*M + cat_new.y).time>cat_new.cat_time ||A.get(cat_new.x*M + cat_new.y).time==-1) && A.get(cat_new.x*M + cat_new.y).visited){
					A.get(cat_new.x*M + cat_new.y).cat_time=cat_time;
					A.get(cat_new.x*M + cat_new.y).visited = false;
					A.get(cat_new.x*M + cat_new.y).path = A.get((cat_new.x+1)*M + cat_new.y).path+"U";
					queue.add(A.get(cat_new.x*M + cat_new.y));
				}
				cat_new.x++;
			}

		}
		if(A.get(cat.x*M + cat.y).time==-1) System.out.println("infinity");
		else System.out.println(A.get(cat.x*M + cat.y).time-1);
		if(original_cat.x==cat.x && original_cat.y==cat.y) System.out.println("stay");
		else System.out.println(A.get(cat.x*M + cat.y).path);
		
	}
}


